ec/src/models/bw6/mod.rs

```diff
@@ -185,113 +211,97 @@ impl<P: BW6Config> BW6<P> {
-        let mut tmp4_p3 = f6p_8p;
-        tmp4_p3.cyclotomic_inverse_in_place();
-        let result11 = result10 * &f5_7p * &f2p * &tmp4_p3;
-
-        // step 11
-        let result12 = result11.square();
-        let f3_6 = f3 * &f6;
-        let f1_7 = f1 * &f7;
-        let mut tmp5_p3 = f1_7 * &f2;
-        tmp5_p3.cyclotomic_inverse_in_place();
-        let result13 = result12 * &f3_6 * &f9p * &tmp5_p3;
-
-        // step 12
-        let result14 = result13.square();
-        let mut tmp6_p3 = f4_2p * &f5_7p * &f6p_8p;
-        tmp6_p3.cyclotomic_inverse_in_place();
-        let result15 = result14 * &f0 * &f0p * &f3p * &f5p * &tmp6_p3;
-
-        // step 13
-        let result16 = result15.square();
-        let mut tmp7_p3 = f3_6;
-        tmp7_p3.cyclotomic_inverse_in_place();
-        let result17 = result16 * &f1p * &tmp7_p3;
-
-        // step 14
-        let result18 = result17.square();
-        let mut tmp8_p3 = f2_4p * &f4_2p_5p * &f9p;
-        tmp8_p3.cyclotomic_inverse_in_place();
-        let result19 = result18 * &f1_7 * &f5_7p * &f0p * &tmp8_p3;
-
-        result19
+    fn final_exponentiation_hard_part(f: &Fp6<P::Fp6Config>) -> Fp6<P::Fp6Config> {
+        // Generic implementation of the hard part of the final exponentiation for the BW6 family.
+        // Computes (u+1)*Phi_k(p(u))/r(u)
+        if P::T_MOD_R_IS_ZERO {
+            // Algorithm 4.3 from https://yelhousni.github.io/phd.pdf
+            // Follows the implementation https://gitlab.inria.fr/zk-curves/snark-2-chains/-/blob/master/sage/pairing_bw6_bls12.py#L1036
+
+            // A = m**(u-1)
+            let a = Self::exp_by_x_minus_1(f);
+            // A = A**(u-1)
+            let a = Self::exp_by_x_minus_1(&a);
+            // A = (m * A).conjugate() * m.frobenius()
+            let a = (f * &a).cyclotomic_inverse().unwrap() * f.frobenius_map(1);
+            // B = A**(u+1) * m
+            let b = Self::exp_by_x_plus_1(&a) * f;
+            // A = A**2 * A
+            let a = a.square() * &a;
+            // A = A.conjugate()
+            let a = a.cyclotomic_inverse().unwrap();
+            // C = B**((u-1)//3)
+            let c = Self::exp_by_x_minus_1_div_3(&b);
+            // D = C**(u-1)
+            let d = Self::exp_by_x_minus_1(&c);
+            // E = (D**(u-1))**(u-1) * D
+            let e = Self::exp_by_x_minus_1(&Self::exp_by_x_minus_1(&d)) * &d;
+            // F = (E**(u+1) * C).conjugate() * D
+            let f = (Self::exp_by_x_plus_1(&e) * &c)
+                .cyclotomic_inverse()
+                .unwrap()
+                * &d;
+            // G = ((F * D)**(u+1)).conjugate() * C * B
+            let g = Self::exp_by_x_plus_1(&(f * &d))
+                .cyclotomic_inverse()
+                .unwrap()
+                * &c
+                * &b;
+            // d2 = (ht**2+3*hy**2)//4
+            let d2 = ((P::H_T * P::H_T + 3 * P::H_Y * P::H_Y) / 4) as u32;
+            let d2 = <P::Fp as PrimeField>::BigInt::from(d2);
```

Comment:
This is fine as we're taking squares. But you can simply by avoiding the explicit conversion to `BigInt`